File: Source/Aladdin.CAPI.Java/src/aladdin/capi/Culture.java
package aladdin.capi;
import aladdin.capi.pbe.*; 
import aladdin.*; 
import aladdin.asn1.*; 
import aladdin.asn1.iso.*; 
import aladdin.asn1.iso.pkcs.*; 
import aladdin.asn1.iso.pkcs.pkcs7.*; 
import java.io.*; 

///////////////////////////////////////////////////////////////////////////
// Национальные особенности
///////////////////////////////////////////////////////////////////////////
public abstract class Culture
{
    // параметры алгоритмов
	public AlgorithmIdentifier hashAlgorithm              (IRand rand) throws IOException { return null; } 
    public AlgorithmIdentifier hmacAlgorithm              (IRand rand) throws IOException { return null; } 
	public AlgorithmIdentifier cipherAlgorithm            (IRand rand) throws IOException { return null; } 
	public AlgorithmIdentifier keyWrapAlgorithm           (IRand rand) throws IOException { return null; } 
	public AlgorithmIdentifier ciphermentAlgorithm        (IRand rand) throws IOException { return null; } 
    public AlgorithmIdentifier signHashAlgorithm          (IRand rand) throws IOException { return null; } 
    public AlgorithmIdentifier signDataAlgorithm          (IRand rand) throws IOException { return null; } 
	public AlgorithmIdentifier transportKeyAlgorithm      (IRand rand) throws IOException { return null; }
	public AlgorithmIdentifier transportAgreementAlgorithm(IRand rand) throws IOException { return null; } 
    
    // параметры шифрования по паролю
    public PBECulture pbe(PBEParameters parameters)
    {
        // параметры шифрования по паролю
        return new PBEDefaultCulture(this, parameters, true); 
    }
	///////////////////////////////////////////////////////////////////////
	// Параметры шифрования на открытом ключе (KeyX)
	///////////////////////////////////////////////////////////////////////
    public AlgorithmIdentifier keyxParameters(
        Factory factory, SecurityStore scope, IRand rand, KeyUsage keyUsage) throws IOException
    {
        // при допустимости транспорта ключа
        if (keyUsage.contains(KeyUsage.KEY_ENCIPHERMENT))
        {
            // создать список параметров 
            AlgorithmIdentifier keyxParameters; 
            
            // получить параметры алгоритма обмена
            if ((keyxParameters = transportKeyAlgorithm(rand)) != null)
            try {
                // получить алгоритм транспорта ключа
                IAlgorithm algorithm = factory.createAlgorithm(
                    scope, keyxParameters, TransportKeyWrap.class
                );
                // проверить наличие алгоритма транспорта ключа
                if (algorithm != null) { RefObject.release(algorithm); return keyxParameters; }
            }
            // обработать возможную ошибку
            catch (IOException e) {}
        }
        // при допустимости согласования ключа ключа
        if (keyUsage.containsAny(KeyUsage.KEY_AGREEMENT))
        {
            // создать список параметров 
            AlgorithmIdentifier keyxParameters; 
            
            // получить параметры алгоритма обмена
            if ((keyxParameters = transportAgreementAlgorithm(rand)) != null)
            try {
                // получить алгоритм согласования ключа
                IAlgorithm algorithm = factory.createAlgorithm(
                    scope, keyxParameters, ITransportAgreement.class
                );
                // проверить наличие алгоритма транспорта ключа
                if (algorithm != null) { RefObject.release(algorithm); return keyxParameters; }
            }
            // обработать возможную ошибку
            catch (IOException e) {}
        }
        return null; 
    }
	///////////////////////////////////////////////////////////////////////
	// Зашифровать данные на открытом ключе (KeyX)
	///////////////////////////////////////////////////////////////////////
    public static ContentInfo keyxEncryptData(Culture culture, 
        Factory factory, SecurityStore scope, IRand rand, 
        Certificate recipientCertificate, CMSData data, 
        Attributes attributes) throws IOException
	{
        // зашифровать данные
        return Culture.keyxEncryptData(culture, factory, scope, rand, 
            recipientCertificate, recipientCertificate.keyUsage(), data, attributes
        ); 
    }
    public static ContentInfo keyxEncryptData(Culture culture, 
        Factory factory, SecurityStore scope, IRand rand, 
        Certificate recipientCertificate, KeyUsage recipientUsage, 
        CMSData data, Attributes attributes) throws IOException
	{
        // зашифровать данные
        return Culture.keyxEncryptData(culture, factory, scope,
            rand, new Certificate[] { recipientCertificate }, 
            new KeyUsage[] { recipientUsage }, 
            new Culture[] { culture }, data, attributes
        ); 
    }
	public static ContentInfo keyxEncryptData(
        Culture culture, Factory factory, SecurityStore scope, 
        IRand rand, Certificate[] recipientCertificates, 
        Culture[] cultures, CMSData data, Attributes attributes) throws IOException
    {
        // создать список использований ключа
        KeyUsage[] recipientUsages = new KeyUsage[recipientCertificates.length]; 

        // заполнить список использования ключа
        for (int i = 0; i < recipientCertificates.length; i++)
        {
            // указать использование ключа
            recipientUsages[i] = recipientCertificates[i].keyUsage(); 
        }
        // зашифровать данные
        return keyxEncryptData(culture, factory, scope, rand, 
            recipientCertificates, recipientUsages, cultures, data, attributes
        ); 
    }
	public static ContentInfo keyxEncryptData(
        Culture culture, Factory factory, SecurityStore scope, IRand rand, 
        Certificate[] recipientCertificates, KeyUsage[] recipientUsages, 
        Culture[] cultures, CMSData data, Attributes attributes) throws IOException
	{
        // проверить указание алгоритмов
        if (culture == null) throw new UnsupportedOperationException(); 
        
        // указать идентификатор типа 
        ObjectIdentifier dataType = new ObjectIdentifier(
            aladdin.asn1.iso.pkcs.pkcs7.OID.ENVELOPED_DATA
        ); 
        // создать список параметров 
        AlgorithmIdentifier[] keyxParameters = 
            new AlgorithmIdentifier[recipientCertificates.length]; 
        
        // указать алгоритм шифрования
        AlgorithmIdentifier cipherParameters = culture.cipherAlgorithm(rand); 
            
        // проверить указание алгоритма
        if (cipherParameters == null) throw new UnsupportedOperationException(); 

        // для всех сертификатов
        for (int i = 0; i < keyxParameters.length; i++)
        {
            // указать используемую культуру
            Culture keyxCulture = (cultures != null) ? cultures[i] : culture; 
            
            // получить параметры алгоритма
            keyxParameters[i] = keyxCulture.keyxParameters(factory, scope, rand, recipientUsages[i]); 
            
            // проверить отсутствие ошибок
            if (keyxParameters[i] == null) throw new UnsupportedOperationException(); 
        }
        // зашифровать данные
        EnvelopedData envelopedData = CMS.keyxEncryptData(
            factory, scope, rand, recipientCertificates, recipientUsages, 
            keyxParameters, cipherParameters, data, attributes
        ); 
        // вернуть закодированную структуру
        return new ContentInfo(dataType, envelopedData); 
    }
	public static ContentInfo keyxEncryptData(
        Culture culture, IRand rand, IPrivateKey privateKey, 
        Certificate[] certificateChain, Certificate recipientCertificate, 
        CMSData data, Attributes attributes) throws IOException
    {
        // зашифровать данные
        return keyxEncryptData(culture, rand, privateKey, certificateChain, 
            recipientCertificate, recipientCertificate.keyUsage(), data, attributes
        ); 
    }
	public static ContentInfo keyxEncryptData(
        Culture culture, IRand rand, IPrivateKey privateKey, 
        Certificate[] certificateChain, Certificate recipientCertificate, 
        KeyUsage recipientUsage, CMSData data, Attributes attributes) throws IOException
    {
        // зашифровать данные
        return keyxEncryptData(culture, rand, privateKey, certificateChain, 
            new Certificate[] { recipientCertificate }, new KeyUsage[] { recipientUsage }, 
            new Culture[] { culture }, data, attributes
        ); 
    }
	public static ContentInfo keyxEncryptData(Culture culture, IRand rand, 
        IPrivateKey privateKey, Certificate[] certificateChain, 
        Certificate[] recipientCertificates, Culture[] cultures, 
        CMSData data, Attributes attributes) throws IOException
    {
        // создать список использований ключа
        KeyUsage[] recipientUsages = new KeyUsage[recipientCertificates.length]; 

        // заполнить список использования ключа
        for (int i = 0; i < recipientCertificates.length; i++)
        {
            // указать использование ключа
            recipientUsages[i] = recipientCertificates[i].keyUsage(); 
        }
        // зашифровать данные
        return keyxEncryptData(culture, rand, privateKey, certificateChain, 
            recipientCertificates, recipientUsages, cultures, data, attributes
        ); 
    }
	public static ContentInfo keyxEncryptData(Culture culture, IRand rand, 
        IPrivateKey privateKey, Certificate[] certificateChain, 
        Certificate[] recipientCertificates, KeyUsage[] recipientUsages, 
        Culture[] cultures, CMSData data, Attributes attributes) throws IOException
	{
        // указать идентификатор типа 
        ObjectIdentifier dataType = new ObjectIdentifier(
            aladdin.asn1.iso.pkcs.pkcs7.OID.ENVELOPED_DATA
        ); 
        // создать список параметров 
        AlgorithmIdentifier[] keyxParameters = 
            new AlgorithmIdentifier[recipientCertificates.length]; 

        // получить параметры алгоритма шифрования
        AlgorithmIdentifier cipherParameters = culture.cipherAlgorithm(rand);
            
        // проверить указание алгоритма
        if (cipherParameters == null) throw new UnsupportedOperationException(); 
            
        // для всех сертификатов
        for (int i = 0; i < keyxParameters.length; i++)
        {
            // указать используемую культуру
            Culture keyxCulture = (cultures != null) ? cultures[i] : culture; 
                
            // получить параметры алгоритма
            keyxParameters[i] = keyxCulture.keyxParameters( 
                privateKey.factory(), privateKey.scope(), rand, recipientUsages[i]
            ); 
            // проверить отсутствие ошибок
            if (keyxParameters[i] == null) throw new UnsupportedOperationException(); 
        }
        // зашифровать данные
        EnvelopedData envelopedData = CMS.keyxEncryptData( 
            rand, privateKey, certificateChain, recipientCertificates, 
            recipientUsages, keyxParameters, cipherParameters, data, attributes
        ); 
        // вернуть закодированную структуру
        return new ContentInfo(dataType, envelopedData); 
    }
	///////////////////////////////////////////////////////////////////////
	// Подписать данные на личном ключе
    // Национальные особенности выбираются согласно сертификату
	///////////////////////////////////////////////////////////////////////
	public static ContentInfo signData(Culture culture, IRand rand, 
        IPrivateKey privateKey, Certificate[] certificateChain, CMSData data, 
        Attributes[] authAttributes, Attributes[] unauthAttributes) throws IOException
    {
        // подписать данные
        return Culture.signData(new Culture[] { culture }, rand, 
            new IPrivateKey[] { privateKey }, new Certificate[][] { certificateChain }, 
            data, authAttributes, unauthAttributes
        ); 
    }
	public static ContentInfo signData(Culture[] cultures, IRand rand, 
        IPrivateKey[] privateKeys, Certificate[][] certificatesChains, CMSData data, 
        Attributes[] authAttributes, Attributes[] unauthAttributes) throws IOException
	{
        // указать идентификатор типа 
        ObjectIdentifier dataType = new ObjectIdentifier(
            aladdin.asn1.iso.pkcs.pkcs7.OID.SIGNED_DATA
        ); 
	    // создать список параметров хэширования
	    AlgorithmIdentifier[] hashParameters = 
            new AlgorithmIdentifier[privateKeys.length];

        // создать список параметров подписи
	    AlgorithmIdentifier[] signHashParameters = 
            new AlgorithmIdentifier[privateKeys.length];

        // для всех личных ключей
        for (int i = 0; i < privateKeys.length; i++)
        { 
            // получить параметры алгоритма хэширования
            hashParameters[i] = cultures[i].hashAlgorithm(rand);
                
            // получить параметры алгоритма подписи
            signHashParameters[i] = cultures[i].signHashAlgorithm(rand); 
                
            // проверить наличие параметров
            if (hashParameters[i] == null || signHashParameters[i] == null)
            {
                // при ошибке выбросить исключение
                throw new UnsupportedOperationException(); 
            }
        }
        // подписать данные
        SignedData signedData = CMS.signData(rand, 
            privateKeys, certificatesChains, hashParameters, 
            signHashParameters, data, authAttributes, unauthAttributes
        ); 
        // вернуть закодированную структуру
        return new ContentInfo(dataType, signedData); 
	}
}

File: Source/Aladdin.CAPI/CAPI/Culture.cs
using System;

namespace Aladdin.CAPI
{
	///////////////////////////////////////////////////////////////////////////
	// Национальные особенности
	///////////////////////////////////////////////////////////////////////////
	public abstract class Culture
	{
		// параметры алгоритмов
        public virtual ASN1.ISO.AlgorithmIdentifier HashAlgorithm               (IRand rand) { return null; }
        public virtual ASN1.ISO.AlgorithmIdentifier HMacAlgorithm               (IRand rand) { return null; }
		public virtual ASN1.ISO.AlgorithmIdentifier CipherAlgorithm             (IRand rand) { return null; }
		public virtual ASN1.ISO.AlgorithmIdentifier KeyWrapAlgorithm            (IRand rand) { return null; }
		public virtual ASN1.ISO.AlgorithmIdentifier CiphermentAlgorithm         (IRand rand) { return null; }
	    public virtual ASN1.ISO.AlgorithmIdentifier SignHashAlgorithm           (IRand rand) { return null; }
	    public virtual ASN1.ISO.AlgorithmIdentifier SignDataAlgorithm           (IRand rand) { return null; }
		public virtual ASN1.ISO.AlgorithmIdentifier TransportKeyAlgorithm       (IRand rand) { return null; }
		public virtual ASN1.ISO.AlgorithmIdentifier TransportAgreementAlgorithm (IRand rand) { return null; }

        // параметры шифрования по паролю
        public virtual PBE.PBECulture PBE(PBE.PBEParameters parameters)
        {
            // параметры шифрования по паролю
            return new PBE.PBEDefaultCulture(this, parameters, true); 
        }
	    ///////////////////////////////////////////////////////////////////////
	    // Параметры шифрования на открытом ключе (KeyX)
	    ///////////////////////////////////////////////////////////////////////
        public ASN1.ISO.AlgorithmIdentifier KeyxParameters(
            Factory factory, SecurityStore scope, IRand rand, KeyUsage keyUsage)
        {
            // при допустимости транспорта ключа
            if ((keyUsage & KeyUsage.KeyEncipherment) != KeyUsage.None)
            {
                // создать список параметров 
                ASN1.ISO.AlgorithmIdentifier keyxParameters = null; 

                // получить параметры алгоритма обмена
                if ((keyxParameters = TransportKeyAlgorithm(rand)) != null)
                try { 
                    // получить алгоритм транспорта ключа
                    IAlgorithm algorithm = factory.CreateAlgorithm<TransportKeyWrap>(scope, keyxParameters);

                    // проверить наличие алгоритма транспорта ключа
                    if (algorithm != null) { RefObject.Release(algorithm); return keyxParameters; } 
                }
                catch {}
            }
            // при допустимости согласования ключа
            if ((keyUsage & KeyUsage.KeyAgreement) != KeyUsage.None)
            {
                // создать список параметров 
                ASN1.ISO.AlgorithmIdentifier keyxParameters = null; 

                // получить параметры алгоритма согласования
                if ((keyxParameters = TransportAgreementAlgorithm(rand)) != null)
                try { 
                    // получить алгоритм согласования ключа
                    IAlgorithm algorithm = factory.CreateAlgorithm<ITransportAgreement>(scope, keyxParameters);

                    // проверить наличие алгоритма транспорта ключа
                    if (algorithm != null) { RefObject.Release(algorithm); return keyxParameters; }
                }
                catch {}
            }
            return null; 
        }
		///////////////////////////////////////////////////////////////////////
		// Зашифровать данные на открытом ключе (KeyX)
		///////////////////////////////////////////////////////////////////////
		public static ASN1.ISO.PKCS.ContentInfo KeyxEncryptData(Culture culture, 
            Factory factory, SecurityStore scope, IRand rand, 
            Certificate recipientCertificate, CMSData data, ASN1.ISO.Attributes attributes)
		{
            // зашифровать данные
            return Culture.KeyxEncryptData(culture, factory, scope, rand, 
                new Certificate[] { recipientCertificate }, 
                new Culture[] { culture }, data, attributes
            ); 
        }
		public static ASN1.ISO.PKCS.ContentInfo KeyxEncryptData(Culture culture, 
            Factory factory, SecurityStore scope, IRand rand, 
            Certificate[] recipientCertificates, Culture[] cultures,
            CMSData data, ASN1.ISO.Attributes attributes)
		{
            // указать идентификатор типа 
            ASN1.ObjectIdentifier dataType = new ASN1.ObjectIdentifier(
                ASN1.ISO.PKCS.PKCS7.OID.envelopedData
            ); 
            // создать список параметров 
            ASN1.ISO.AlgorithmIdentifier[] keyxParameters = 
                new ASN1.ISO.AlgorithmIdentifier[recipientCertificates.Length]; 

            // указать алгоритм шифрования
            ASN1.ISO.AlgorithmIdentifier cipherParameters = culture.CipherAlgorithm(rand); 

            // проверить указание параметров
            if (cipherParameters == null) throw new NotSupportedException(); 

            // для всех сертификатов
            for (int i = 0; i < keyxParameters.Length; i++)
            {
                // указать используемую культуру
                Culture keyxCulture = (cultures != null) ? cultures[i] : culture; 

                // получить параметры алгоритма
                keyxParameters[i] = keyxCulture.KeyxParameters(
                    factory, scope, rand, recipientCertificates[i].KeyUsage
                ); 
                // проверить отсутствие ошибок
                if (keyxParameters[i] == null) throw new NotSupportedException(); 
            }
            // зашифровать данные
            ASN1.ISO.PKCS.PKCS7.EnvelopedData envelopedData = CMS.KeyxEncryptData(
                factory, scope, rand, recipientCertificates, 
                cipherParameters, keyxParameters, data, attributes
            ); 
            // вернуть закодированную структуру
            return new ASN1.ISO.PKCS.ContentInfo(dataType, envelopedData); 
        }
	    public static ASN1.ISO.PKCS.ContentInfo KeyxEncryptData(
            Culture culture, IRand rand, IPrivateKey privateKey, 
            Certificate[] certificateChain, Certificate recipientCertificate, 
            CMSData data, ASN1.ISO.Attributes attributes)
        {
            // зашифровать данные
            return KeyxEncryptData(culture, rand, privateKey, certificateChain, 
                new Certificate[] { recipientCertificate }, 
                new Culture[] { culture }, data, attributes
            ); 
        }
	    public static ASN1.ISO.PKCS.ContentInfo KeyxEncryptData(
            Culture culture, IRand rand, IPrivateKey privateKey, 
            Certificate[] certificateChain, Certificate[] recipientCertificates, 
            Culture[] cultures, CMSData data, ASN1.ISO.Attributes attributes) 
	    {
            // указать идентификатор типа 
            ASN1.ObjectIdentifier dataType = new ASN1.ObjectIdentifier(
                ASN1.ISO.PKCS.PKCS7.OID.envelopedData
            ); 
            // создать список параметров 
            ASN1.ISO.AlgorithmIdentifier[] keyxParameters = 
                new ASN1.ISO.AlgorithmIdentifier[recipientCertificates.Length]; 

            // получить параметры алгоритма шифрования
            ASN1.ISO.AlgorithmIdentifier cipherParameters = culture.CipherAlgorithm(rand);

            // проверить указание параметров
            if (cipherParameters == null) throw new NotSupportedException(); 
            
            // для всех сертификатов
            for (int i = 0; i < keyxParameters.Length; i++)
            {
                // указать используемую культуру
                Culture keyxCulture = (cultures != null) ? cultures[i] : culture; 

                // получить параметры алгоритма
                keyxParameters[i] = keyxCulture.KeyxParameters( 
                    privateKey.Factory, privateKey.Scope, 
                    rand, recipientCertificates[i].KeyUsage
                ); 
                // проверить отсутствие ошибок
                if (keyxParameters[i] == null) throw new NotSupportedException(); 
            }
            // зашифровать данные
            ASN1.ISO.PKCS.PKCS7.EnvelopedData envelopedData = CMS.KeyxEncryptData( 
                rand, privateKey, certificateChain, recipientCertificates, 
                keyxParameters, cipherParameters, data, attributes
            ); 
            // вернуть закодированную структуру
            return new ASN1.ISO.PKCS.ContentInfo(dataType, envelopedData); 
        }
		///////////////////////////////////////////////////////////////////////
		// Подписать данные на личном ключе
		///////////////////////////////////////////////////////////////////////
		public static ASN1.ISO.PKCS.ContentInfo SignData(Culture[] cultures, IRand rand, 
            IPrivateKey[] privateKeys, Certificate[][] certificatesChains, CMSData data, 
            ASN1.ISO.Attributes[] authAttributes, ASN1.ISO.Attributes[] unauthAttributes)
		{
            // скорректировать переданные параметры
            if (authAttributes   == null) authAttributes   = new ASN1.ISO.Attributes[privateKeys.Length];
            if (unauthAttributes == null) unauthAttributes = new ASN1.ISO.Attributes[privateKeys.Length];

            // указать идентификатор типа 
            ASN1.ObjectIdentifier dataType = new ASN1.ObjectIdentifier(
                ASN1.ISO.PKCS.PKCS7.OID.signedData
            ); 
	        // создать список параметров хэширования
	        ASN1.ISO.AlgorithmIdentifier[] hashParameters = 
                new ASN1.ISO.AlgorithmIdentifier[privateKeys.Length];

            // создать список параметров подписи
	        ASN1.ISO.AlgorithmIdentifier[] signHashParameters = 
                new ASN1.ISO.AlgorithmIdentifier[privateKeys.Length];

            // для всех личных ключей
            for (int i = 0; i < privateKeys.Length; i++)
            { 
                // получить параметры алгоритма хэширования
                hashParameters[i] = cultures[i].HashAlgorithm(rand); 

                // получить параметры алгоритма подписи
                signHashParameters[i] = cultures[i].SignHashAlgorithm(rand);

                // проверить наличие параметров
                if (hashParameters[i] == null || signHashParameters[i] == null)
                {
                    // при ошибке выбросить исключение
                    throw new NotSupportedException(); 
                }
            }
            // подписать данные
            ASN1.ISO.PKCS.PKCS7.SignedData signedData = CMS.SignData(
                rand, privateKeys, certificatesChains, hashParameters, 
                signHashParameters, data, authAttributes, unauthAttributes
		    ); 
		    // вернуть закодированную структуру
		    return new ASN1.ISO.PKCS.ContentInfo(dataType, signedData); 
		}
		public static ASN1.ISO.PKCS.ContentInfo SignData(
            Culture culture, IRand rand, IPrivateKey privateKey, Certificate[] certificateChain, 
            CMSData data, ASN1.ISO.Attributes authAttributes, ASN1.ISO.Attributes unauthAttributes)
		{
            // подписать данные
            return Culture.SignData(new Culture[] { culture }, rand, 
                new IPrivateKey[]   { privateKey       }, 
                new Certificate[][] { certificateChain }, data, 
                new ASN1.ISO.Attributes[] { authAttributes   },
                new ASN1.ISO.Attributes[] { unauthAttributes }
            ); 
		}
    }
}
